zenoh/src/net/routing/interceptor/authorization.rs

```diff
@@ -0,0 +1,172 @@
+use ahash::RandomState;
+use std::collections::HashMap;
+use zenoh_config::{
+    AclConfig, Action, ConfigRule, Permission, PolicyRule, Subject, NUMBER_OF_ACTIONS,
+    NUMBER_OF_PERMISSIONS,
+};
+use zenoh_keyexpr::keyexpr;
+use zenoh_keyexpr::keyexpr_tree::{IKeyExprTree, IKeyExprTreeMut, KeBoxTree};
+use zenoh_result::ZResult;
+
+pub struct PolicyForSubject(Vec<Vec<KeTreeRule>>); //vec of actions over vec of permission for tree of ke for this
+pub struct PolicyMap(pub HashMap<i32, PolicyForSubject, RandomState>); //index of subject (i32) instead of subject (String)
+
+type KeTreeRule = KeBoxTree<bool>;
+
+pub struct PolicyEnforcer {
+    pub(crate) acl_enabled: bool,
+    pub(crate) default_permission: Permission,
+    pub(crate) subject_map: Option<HashMap<Subject, i32, RandomState>>,
+    pub(crate) policy_map: Option<PolicyMap>,
+}
+
+#[derive(Debug, Clone)]
+
+pub struct PolicyInformation {
+    subject_map: HashMap<Subject, i32, RandomState>,
+    policy_rules: Vec<PolicyRule>,
+}
+
+impl PolicyEnforcer {
+    pub fn new() -> PolicyEnforcer {
+        PolicyEnforcer {
+            acl_enabled: true,
+            default_permission: Permission::Deny,
+            subject_map: None,
+            policy_map: None,
+        }
+    }
+
+    /*
+       initializes the policy_enforcer
+    */
+    pub fn init(&mut self, acl_config: AclConfig) -> ZResult<()> {
+        self.acl_enabled = acl_config.enabled;
+        self.default_permission = acl_config.default_permission;
+        if self.acl_enabled {
+            if let Some(rules) = acl_config.rules {
+                if rules.is_empty() {
+                    log::warn!("ACL ruleset in config file is empty!!!");
+                    self.policy_map = None;
+                    self.subject_map = None;
+                }
+                let policy_information = self.policy_information_point(rules)?;
+
+                let subject_map = policy_information.subject_map;
+                let mut main_policy: PolicyMap = PolicyMap(HashMap::default());
+                //first initialize the vector of vectors (required to maintain the indices)
+                for index in subject_map.values() {
+                    let mut rule: PolicyForSubject = PolicyForSubject(Vec::new());
+                    for _i in 0..NUMBER_OF_ACTIONS {
```

Comment:
I believe this is a `ill-typed` way of doing it. If the actual index of the policy is harcoded, why not having it expressed in the actual type as different fields?